DirectProgramming/DPC++FPGA/ReferenceDesigns/cholesky/README.md

```diff
@@ -0,0 +1,263 @@
+| Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX)      | 214k matrices/s for real matrices of size 32 * 32
+
+
+## Purpose
+
+This FPGA reference design demonstrates the use of the Cholesky decomposition header library, a common operation employed in linear algebra.
+The Cholesky decomposition takes a hermitian, positive-definite matrix _A_ (input) and computes the lower triangular matrix _L_ such that: _L_ * _L*_ = A, where _L*_ in the conjugate transpose of _L_.
+The algorithms employed by the header library is an FPGA throughput optimized version of the Cholesky–Banachiewicz algorithm. Background information on the Cholesky decomposition and this algorithm can be found in Wikipedia's [Cholesky decomposition](https://https://en.wikipedia.org/wiki/Cholesky_decomposition) article. 
+
+The Cholesky decomposition is used extensively in machine learning applications such as least-squares regressions and Monte-Carlo simulations.
+
+
+### Matrix dimensions and FPGA resources
+
+The Cholesky decomposition algorithm factors a real _n_ × _n_ matrix. The algorithm computes the vector dot product of two rows of the matrix. In our FPGA implementation, the dot product is computed in a loop over the row's _n_ elements. The loop is fully unrolled to maximize throughput. As a result, *n* real multiplication operations are performed in parallel on the FPGA, followed by sequential additions to compute the dot product result.
+
+We use the compiler flag `-fp-relaxed`, which permits the compiler to reorder floating point additions (i.e. to assume that floating point addition is commutative). The compiler uses this freedom to reorder the additions so that the dot product arithmetic can be optimally implemented using the FPGA's specialized floating point DSP (Digital Signal Processing) hardware.
+
+With this optimization, our FPGA implementation requires _n_ DSPs to compute the real floating point dot product.
+The input matrix is also replicated two times in order to be able to read two full rows per cycle.
+Thus, the matrix size is constrained by the total FPGA DSP and RAM resources available.
+
+
+## Key Implementation Details
+| Kernel            | Description
+---                 |---
+| Cholesky          | Implements a modified Cholesky–Banachiewicz Cholesky decomposition algorithm.
+
+To optimize the performance-critical loop in its algorithm, the design leverages concepts discussed in the following FPGA tutorials:
+* **Triangular Loop Optimization** (triangular_loop)
+* **Explicit Pipelining with `fpga_reg`** (fpga_register)
+* **Loop `ivdep` Attribute** (loop_ivdep)
+* **Unrolling Loops** (loop_unroll)
+
+ The key optimization techniques used are as follows:
+   1. Traversing the matrix in a column fashion to increase the read-after-write (RAW) loop iteration distance 
+   2. Using two copies of the compute matrix in order to be able to read two full rows per cycle
```

Comment:
This is done explicitly because the two matrices are stored differently: one is store row-wise while the other is stored column-wise in order to later on read a full row from one and a full column from the other at the same time

Reply:
I've modified this line because saying "to be able to read two full rows per cycle" was incorrect